client/src/controllers/ZoneController.ts

```diff
@@ -53,15 +72,41 @@ export class ZoneController extends Controller {
     this.modeValue = ZoneMode.Inactive;
   }
 
-  modeValueChanged(current: ZoneMode) {
+  modeValueChanged(current: ZoneMode, previous?: ZoneMode) {
+    const element = this.element;
     const activeClasses = this.activeClasses;
+    const inactiveClasses = this.inactiveClasses;
+
+    if (
+      !(activeClasses.length + inactiveClasses.length) ||
+      previous === current
+    )
+      return;
 
-    if (!activeClasses.length) return;
+    if (previous === undefined) {
+      if (
+        activeClasses.every((className) =>
+          element.classList.contains(className),
+        )
+      ) {
+        this.modeValue = ZoneMode.Active;
+      }
+
+      if (
+        inactiveClasses.every((className) =>
+          element.classList.contains(className),
+        )
+      ) {
+        this.modeValue = ZoneMode.Inactive;
+      }
```

Comment:
Let's return here, so that we rely on the method re-running (as `modeValue` may have changed).
I might be nice to add some comments describing what's happening here.

I can do these when merging.

Reply:
@lb- I thought this while writing test cases and it clicked me with your comment that not adding `data-w-zone-inactive-class` attribute when initial state should be active makes mode value inacitve here is what i did to figure out this:

I thought it would jest runtime conflicts where sometimes events trigger later than asserting, so i set up stimulus and mimic ZoneController (if you want to check this is code):

```
import { Controller } from "@hotwired/stimulus";

export default class extends Controller {
  static values = { mode: { type: String, default: "" } };
  static classes = ["inactive", "active"];

  connect() {
    console.log("test-change-callback Connected");
  }

  initialize() {
    console.log("test-change-callback Initialize");
  }

  changeToggleVal() {
    this.modeValue = "inactivate";
  }

  modeValueChanged(current, previous) {
    const element = this.element;
    const activeClasses = this.activeClasses;
    const inactiveClasses = this.inactiveClasses;

    console.log("inside callback");
    if (previous === undefined) {
      if (
        activeClasses.every((className) =>
          element.classList.contains(className)
        )
      ) {
        console.log(element.classList);
        this.modeValue = "active";
      }

      console.log(
        element.classList,
        inactiveClasses,
        inactiveClasses.every((className) =>
          element.classList.contains(className)
        )
      );

      if (
        inactiveClasses.every((className) =>
          element.classList.contains(className))
      ) {
        this.modeValue = "inactive";
      }
      return;
    }

    console.log(`Current: ${current}, Previous: ${previous}`);
    if (current === "active") {
      element.classList.add(...activeClasses);
      element.classList.remove(...inactiveClasses);
    } else {
      element.classList.add(...inactiveClasses);
      element.classList.remove(...activeClasses);
    }
  }
}
```

so the issue with `every` here, it returns true if the array is empty so when not adding inactive stimulus class attribute inactive is now empty array and becomes true and override active state HTML example with our case:

```
<div
          class="mode-zone"
          data-controller="w-zone"
          data-w-zone-active-class="mode-zone"
></div>
```
- No inactive class means [](empty array) means `every ` method will evolute `true` (override modeValue)

### Solution

We can add `return` inside if block of acitveClasses check :

```
if (activeClasses.every((className) => element.classList.contains(className),)) {
        this.modeValue = ZoneMode.Active;
        return;
      }
```
- Now no need to declare `data-w-zone-inactive-class` in HTML explicitly

So final will be like this:

```
if (previous === undefined) {
    if (
      activeClasses.every((className) =>
        element.classList.contains(className),
      )
    ) {
      this.modeValue = ZoneMode.Active;
      return;
    }

    if (
      inactiveClasses.every((className) =>
        element.classList.contains(className),
      )
    ) {
      this.modeValue = ZoneMode.Inactive;
      return;
    }
  }
```

`return` on `previous === undefined` condition block will prevent to add inactive classes:

```
<div
          class=""
          data-controller="w-zone"
          data-w-zone-inactive-class="w-hidden"
        ></div>
```
- Default mode value is `ZoneMode.Inactive` if no mode value is provided via HTML, so in above case inactive class will not be added because `previous ` is undefined and in that condition we returned modeValueChanged method.

#### Sorry for not mentioning this scenario after adding last unit test.

Reply:
i have added commit for this on this branch df26a3f.

Reply:
Great, I have tested this also and I think it makes sense. Thanks for thinking it through.

Please can you prepare a new PR with just this commit @ayaan-qadri ping me on the PR and I will do a review.